File: backend/app/schemas/industry.py
from pydantic import BaseModel, Field
from typing import Optional, List, Any
from datetime import datetime

from .industry_outcomes import OutcomeOverlay


class IndustryRequirementResponse(BaseModel):
    id: str
    role: str
    skill: str
    skill_slug: Optional[str] = None
    skill_category: str
    required_level: float = Field(default=0.75, ge=0.0, le=1.0)
    importance: float = Field(..., ge=0.0, le=1.0)
    demand: float = Field(..., ge=0.0, le=1.0)
    interview_relevance: float = Field(..., ge=0.0, le=1.0)
    industry_confidence: float = Field(default=0.85, ge=0.0, le=1.0)
    source: str
    source_url: Optional[str] = None
    source_version: Optional[str] = None
    source_occupation: Optional[str] = None
    source_reference: Optional[str] = None
    mapping_version: Optional[str] = None
    role_relevance: Optional[str] = None
    source_concept: Optional[str] = None
    canonical_mapping: Optional[str] = None
    mapping_rationale: Optional[str] = None
    source_quality: float = Field(default=0.85, ge=0.0, le=1.0)
    evidence_context: Optional[str] = None
    description: Optional[str] = None
    version: str = "2026.1"
    metadata: Optional[dict] = None
    published_at: Optional[str] = None
    retrieved_at: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    # Phase 7: requirement trustworthiness (all optional for compatibility
    # with rows persisted before these fields existed).
    evidence_strength: Optional[str] = Field(
        default=None,
        description="Trust tier: strong, moderate, weak, or insufficient",
    )
    supporting_chunks: Optional[List[dict]] = Field(
        default=None,
        description="References (chunk id, role, topic, source, url) to benchmark chunks mentioning this skill",
    )
    duplicate_sources_collapsed: Optional[int] = Field(
        default=None,
        description="Same-source duplicate rows collapsed during aggregation",
    )
    # Phase 3: contextual employer-outcome annotation. Absent unless the
    # outcome overlay is enabled and the cell meets its threshold. Never a
    # gap-mathematics input.
    outcome_overlay: Optional[OutcomeOverlay] = Field(
        default=None,
        description="Observed employer-demand context for this curated requirement",
    )
    # Dynamic Industry Intelligence (additive, all optional for compatibility).
    location: Optional[dict] = Field(
        default=None,
        description="Location scope {scope, country, region, city, label, is_global}; None means global/default",
    )
    trend: Optional[str] = Field(
        default=None,
        description="Skill trend: stable, rising, emerging, declining, insufficient_data",
    )
    freshness: Optional[str] = Field(
        default=None,
        description="Freshness state: current, aging, stale, unknown",
    )
    data_origin: Optional[str] = Field(
        default=None,
        description="Provenance tier: source_data, inaura_derived, demo_seeded",
    )
    data_origins: Optional[List[str]] = Field(default=None)
    collected_at: Optional[str] = Field(default=None)
    last_updated: Optional[str] = Field(default=None)
    location_match: Optional[str] = Field(
        default=None,
        description="How the row relates to the requested location: global, exact, fallback_global",
    )
    mapping_status: Optional[str] = Field(
        default=None,
        description="Canonical mapping state: mapped or unmapped (unmapped preserves source concept)",
    )

    class Config:
        from_attributes = True


class RoleSummaryResponse(BaseModel):
    title: str
    slug: str
    category: str
    description: str
    aliases: List[str]
    source_benchmarks: List[str]


class SharedSkillItem(BaseModel):
    skill: str
    category: str
    required_level_a: float
    required_level_b: float
    importance_a: float
    importance_b: float
    level_delta: float
    importance_delta: float


class UniqueSkillItem(BaseModel):
    skill: str
    category: str
    required_level: float
    importance: float


class RoleComparisonResponse(BaseModel):
    role_a: str
    role_b: str
    overlap_score: float
    transition_effort: str
    shared_skill_count: int
    unique_to_a_count: int
    unique_to_b_count: int
    shared_skills: List[SharedSkillItem]
    unique_skills_a: List[UniqueSkillItem]
    unique_skills_b: List[UniqueSkillItem]
    transferable_skills: List[str]
    key_skills_to_acquire: List[str]
    transition_advice: str


class CustomRoleRequest(BaseModel):
    role: str = Field(..., min_length=2, max_length=100, description="Custom role title, e.g., AI Engineer")
    query: Optional[str] = Field(None, max_length=400, description="Optional focus query")


class CustomRoleResponse(BaseModel):
    status: str
    role: str
    is_custom: bool
    requirements: List[IndustryRequirementResponse]
    skills_count: int
    confidence: float
    source_summary: str
    note: str


class RetrieveRequest(BaseModel):
    role: str = Field(..., min_length=2, max_length=100, description="Target role, e.g., Software Engineer")
    query: Optional[str] = Field(None, max_length=400, description="Optional free-text query; defaults to role")
    top_k: int = Field(10, ge=1, le=30, description="Number of requirements to retrieve")
    location: Optional[str] = Field(None, max_length=120, description="Optional location label, e.g., Bengaluru; omit for global")


class RetrieveItem(BaseModel):
    id: str
    role: str
    skill: str
    skill_category: str
    required_level: float = 0.75
    importance: float
    demand: float
    interview_relevance: float
    industry_confidence: float = 0.85
    source: str
    source_url: Optional[str] = None
    source_version: Optional[str] = None
    source_occupation: Optional[str] = None
    source_reference: Optional[str] = None
    mapping_version: Optional[str] = None
    role_relevance: Optional[str] = None
    source_concept: Optional[str] = None
    canonical_mapping: Optional[str] = None
    mapping_rationale: Optional[str] = None
    source_quality: float = 0.85
    evidence_context: Optional[str] = None
    description: Optional[str] = None
    version: str = "2026.1"
    similarity: float = Field(..., description="Retrieval relevance 0-1, not proficiency")
    evidence_strength: Optional[str] = Field(
        default=None,
        description="Trust tier: strong, moderate, weak, or insufficient",
    )
    supporting_chunks: Optional[List[dict]] = Field(
        default=None,
        description="References (chunk id, role, topic, source, url) to benchmark chunks mentioning this skill",
    )
    # Phase 3: contextual employer-outcome annotation (see IndustryRequirementResponse).
    outcome_overlay: Optional[OutcomeOverlay] = Field(
        default=None,
        description="Observed employer-demand context for this curated requirement",
    )
    trend: Optional[str] = Field(default=None)
    freshness: Optional[str] = Field(default=None)
    data_origin: Optional[str] = Field(default=None)
    location: Optional[dict] = Field(default=None)
    collected_at: Optional[str] = Field(default=None)
    last_updated: Optional[str] = Field(default=None)


class RetrieveResponse(BaseModel):
    role: str
    query: str
    count: int
    items: List[RetrieveItem]
    note: str = Field(default="INAURA Industry Knowledge Catalog — authentic benchmarks with source attribution")


class IndustryIntelligenceSkill(BaseModel):
    id: Optional[str] = None
    role: str
    skill: str
    skill_slug: Optional[str] = None
    skill_category: Optional[str] = "General"
    required_level: float = 0.75
    importance: float = 0.5
    demand: float = 0.5
    interview_relevance: float = 0.5
    industry_confidence: float = 0.85
    trend: Optional[str] = None
    freshness: Optional[str] = None
    source: str = ""
    source_url: Optional[str] = None
    evidence_context: Optional[str] = None
    data_origin: Optional[str] = None
    data_origins: Optional[List[str]] = None
    location: Optional[dict] = None
    location_match: Optional[str] = None
    mapping_status: Optional[str] = None
    source_concept: Optional[str] = None
    canonical_mapping: Optional[str] = None
    collected_at: Optional[str] = None
    last_updated: Optional[str] = None
    published_at: Optional[str] = None
    retrieved_at: Optional[str] = None


class IndustryIntelligenceResponse(BaseModel):
    role: str
    location: dict
    location_match: str = "global"
    last_updated: Optional[str] = None
    skills: List[IndustryIntelligenceSkill]
    counts: Optional[dict] = None
    data_origins: Optional[List[str]] = None
    note: str = ""


class DemandProviderInfo(BaseModel):
    provider_id: str
    display_name: str
    data_origin: str
    is_live: bool
    description: str

File: frontend/src/services/industry.ts
import { apiFetch } from "./api";

export type OutcomeOverlay = {
  observed_demand: number | null;
  observed_skill_gap: number | null;
  outcome_sample: { requirements_n: number; applied_n: number; feedback_count: number };
  outcome_window: { window_from: string | null; window_to: string | null; basis: string };
  outcome_provenance: {
    signal_version: string;
    scope: string;
    location: string;
    role: string | null;
    skill: string | null;
  };
  stale: boolean;
};

export type EmergingSkill = {
  skill: string;
  skill_id: string | null;
  observed_demand: number | null;
  observed_skill_gap: number | null;
  outcome_sample: OutcomeOverlay["outcome_sample"];
  outcome_window: OutcomeOverlay["outcome_window"];
  status: "observed_not_required";
  role: string | null;
  location: string | null;
};

export type IndustryRequirement = {
  id: string;
  role: string;
  skill: string;
  skill_category: string;
  importance: number;
  demand: number;
  interview_relevance: number;
  source: string;
  source_url: string | null;
  description: string | null;
  version: string;
  metadata: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
  // Phase 3: contextual employer-outcome annotation. Absent unless the
  // overlay is enabled and the cell meets its threshold. Never a math input.
  outcome_overlay?: OutcomeOverlay | null;
  // Dynamic Industry Intelligence (all optional for compatibility)
  required_level?: number;
  industry_confidence?: number;
  trend?: string | null;
  freshness?: string | null;
  data_origin?: string | null;
  location?: Record<string, unknown> | null;
  collected_at?: string | null;
  last_updated?: string | null;
};

export type RetrieveItem = {
  id: string;
  role: string;
  skill: string;
  skill_category: string;
  importance: number;
  demand: number;
  interview_relevance: number;
  source: string;
  source_url: string | null;
  description: string | null;
  version: string;
  similarity: number;
  // Dynamic Industry Intelligence (all optional for compatibility)
  trend?: string | null;
  freshness?: string | null;
  data_origin?: string | null;
  location?: Record<string, unknown> | null;
};

export type RetrieveResponse = {
  role: string;
  query: string;
  count: number;
  items: RetrieveItem[];
  note: string;
};

export type AnalysisState = {
  user_id: string;
  target_role: string | null;
  status: "not_started" | "ready" | "processing" | "completed" | "failed";
  last_retrieval: RetrieveResponse | null;
  created_at: string;
  updated_at: string;
};

export type RoleSummary = {
  title: string;
  slug: string;
  category: string;
  description: string;
  aliases: string[];
  source_benchmarks: string[];
};

/** The catalog roles INAURA has industry benchmarks for, with descriptions */
export function getRolesCatalog() {
  return apiFetch<RoleSummary[]>("/industry/roles/catalog");
}

export function listRoles() {
  return apiFetch<string[]>("/industry/roles");
}

export function getRequirements(role: string) {
  return apiFetch<IndustryRequirement[]>(`/industry/requirements?role=${encodeURIComponent(role)}`);
}

// Phase 3: aggregated employer-observed skills with no curated requirement.
// Read-only; every item carries status "observed_not_required".
export function getEmergingSkills(role: string, location?: string) {
  const q = location
    ? `?role=${encodeURIComponent(role)}&location=${encodeURIComponent(location)}`
    : `?role=${encodeURIComponent(role)}`;
  return apiFetch<EmergingSkill[]>(`/industry/outcomes/emerging${q}`);
}

// export function retrieveIndustry(payload: { role: string; query?: string; top_k?: number }) {
export function retrieveIndustry(payload: { role: string; query?: string; top_k?: number; location?: string }) {
  return apiFetch<RetrieveResponse>("/industry/retrieve", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export type IntelligenceSkill = {
  id?: string | null;
  role: string;
  skill: string;
  skill_slug?: string | null;
  skill_category?: string | null;
  required_level: number;
  importance: number;
  demand: number;
  interview_relevance: number;
  industry_confidence: number;
  trend?: string | null;
  freshness?: string | null;
  source: string;
  source_url?: string | null;
  evidence_context?: string | null;
  data_origin?: string | null;
  data_origins?: string[] | null;
  location?: { label?: string; scope?: string } | null;
  location_match?: string | null;
  mapping_status?: string | null;
  source_concept?: string | null;
  canonical_mapping?: string | null;
  mapping_rationale?: string | null;
  collected_at?: string | null;
  last_updated?: string | null;
  published_at?: string | null;
  retrieved_at?: string | null;
};

export type IntelligenceResponse = {
  role: string;
  location: { label?: string; scope?: string };
  location_match: string;
  last_updated: string | null;
  skills: IntelligenceSkill[];
  counts?: Record<string, number> | null;
  data_origins?: string[] | null;
  note: string;
};

export function getRoleIntelligence(role: string, location?: string) {
  const params = new URLSearchParams({ role });
  if (location && location.trim()) params.set("location", location.trim());
  return apiFetch<IntelligenceResponse>(`/industry/intelligence?${params.toString()}`);
}

export function listDemandProviders() {
  return apiFetch<
    { provider_id: string; display_name: string; data_origin: string; is_live: boolean; description: string }[]
  >("/industry/providers");
}

export function getAnalysisState() {
  return apiFetch<AnalysisState>("/analysis/state");
}

export function setTargetRole(target_role: string) {
  return apiFetch<AnalysisState>("/analysis/target", {
    method: "POST",
    body: JSON.stringify({ target_role }),
  });
}

export function prepareAnalysis(target_role?: string) {
  return apiFetch<{
    status: AnalysisState["status"];
    target_role: string;
    message: string;
    retrieval: RetrieveResponse | null;
  }>("/analysis/prepare", {
    method: "POST",
    body: JSON.stringify(target_role ? { target_role } : {}),
  });
}
